server/core/basic_cluster.go

```diff
@@ -115,195 +102,298 @@ func (bc *BasicCluster) GetFollowerStores(region *RegionInfo) []*StoreInfo {
 	return bc.Regions.GetStorePendingPeerCount(storeID)
 }
 
 // GetStoreWitnessCount gets the total count of a store's witness RegionInfo.
 func (bc *BasicCluster) GetStoreWitnessCount(storeID uint64) int {
-	bc.RLock()
-	defer bc.RUnlock()
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
 	return bc.Regions.GetStoreWitnessCount(storeID)
 }
 
 // GetStoreLeaderRegionSize get total size of store's leader regions.
 func (bc *BasicCluster) GetStoreLeaderRegionSize(storeID uint64) int64 {
-	bc.RLock()
-	defer bc.RUnlock()
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
 	return bc.Regions.GetStoreLeaderRegionSize(storeID)
 }
 
 // GetStoreRegionSize get total size of store's regions.
 func (bc *BasicCluster) GetStoreRegionSize(storeID uint64) int64 {
-	bc.RLock()
-	defer bc.RUnlock()
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
 	return bc.Regions.GetStoreRegionSize(storeID)
 }
 
 // GetAverageRegionSize returns the average region approximate size.
 func (bc *BasicCluster) GetAverageRegionSize() int64 {
-	bc.RLock()
-	defer bc.RUnlock()
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
 	return bc.Regions.GetAverageRegionSize()
 }
 
+// GetRegionByKey searches RegionInfo from regionTree.
+func (bc *BasicCluster) GetRegionByKey(regionKey []byte) *RegionInfo {
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
+	return bc.Regions.GetRegionByKey(regionKey)
+}
+
+// GetPrevRegionByKey searches previous RegionInfo from regionTree.
+func (bc *BasicCluster) GetPrevRegionByKey(regionKey []byte) *RegionInfo {
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
+	return bc.Regions.GetPrevRegionByKey(regionKey)
+}
+
+// ScanRange scans regions intersecting [start key, end key), returns at most
+// `limit` regions. limit <= 0 means no limit.
+func (bc *BasicCluster) ScanRange(startKey, endKey []byte, limit int) []*RegionInfo {
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
+	return bc.Regions.ScanRange(startKey, endKey, limit)
+}
+
+// GetOverlaps returns the regions which are overlapped with the specified region range.
+func (bc *BasicCluster) GetOverlaps(region *RegionInfo) []*RegionInfo {
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
+	return bc.Regions.GetOverlaps(region)
+}
+
+// GetRegionSizeByRange scans regions intersecting [start key, end key), returns the total region size of this range.
+func (bc *BasicCluster) GetRegionSizeByRange(startKey, endKey []byte) int64 {
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
+	return bc.Regions.GetRegionSizeByRange(startKey, endKey)
+}
+
 func (bc *BasicCluster) getWriteRate(
 	f func(storeID uint64) (bytesRate, keysRate float64),
 ) (storeIDs []uint64, bytesRates, keysRates []float64) {
-	bc.RLock()
-	defer bc.RUnlock()
+	bc.Stores.mu.RLock()
+	defer bc.Stores.mu.RUnlock()
 	count := len(bc.Stores.stores)
 	storeIDs = make([]uint64, 0, count)
 	bytesRates = make([]float64, 0, count)
 	keysRates = make([]float64, 0, count)
```

Comment:
it only needs to get all  store ids by using this store read lock. so it can release it earlier